### ISSUE TITLE
AWSSSRWebSocket runLoop Crash

### ISSUE THREAD

**Issue by adamup928** [awslabs/aws-mobile-appsync-sdk-ios#303]
We're seeing crashes among our users with version 2.14.0 of the AppSync SDK.

The crashes are in `AWSSRWebSocket.m` on line 1874, which is just the runloop's while loop.  Maybe the `_runLoop` variable is being deallocated or cleaned up by the OS...can't tell.  A backtrace is below.

I can't find any rhyme or reason to the cause.  Sometimes we're in the foreground, sometimes not.  Sometimes there's no recent activity in the AppSync library.

 - AppSync SDK Version: 2.14.0
 - Dependency Manager: CocoaPods
 - Swift Version : 5
 - Device: iPhone6, iPhone SE, iPhone 8, iPhone 8 Plus, iPhone X, seems all devices affected
 - iOS Version: 12.2, 12.3, 12.4, 12.4.1

```
# Cause: Crash due to signal: SIGSEGV(SEGV_MAPERR) at 00000010

Thread 13 [Crashed]:

0    libobjc.A.dylib                          0x20eabb530     objc_msgSend + 16
1    CoreFoundation                           0x20f86ce6c     _signalEventSync + 216
2    CoreFoundation                           0x20f86c460     _cfstream_shared_signalEventSync + 372
3    CoreFoundation                           0x20f85a2bc     _CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 24
4    CoreFoundation                           0x20f85a23c     _CFRunLoopDoSource0 + 88
5    CoreFoundation                           0x20f859b24     _CFRunLoopDoSources0 + 176
6    CoreFoundation                           0x20f854a60     _CFRunLoopRun + 1004
7    CoreFoundation                           0x20f854354     CFRunLoopRunSpecific + 436
8    Foundation                               0x210221fcc     -[NSRunLoop(NSRunLoop) runMode:beforeDate:] + 300
9    AWSAppSync                               0x1032a1030     -[_SRRunLoopThread main] (AWSSRWebSocket.m:1874)
10   Foundation                               0x21034e6e4     _NSThread__start__ + 984
11   libsystem_pthread.dylib                  0x20f4e82c0     _pthread_body + 128
12   libsystem_pthread.dylib                  0x20f4e8220     _pthread_start + 44
13   libsystem_pthread.dylib                  0x20f4ebcdc     thread_start + 4
```

**Comment by palpatim** [comment]
@adamup928

As noted in #184, we just released [AppSync 3.0.0](https://github.com/awslabs/aws-mobile-appsync-sdk-ios/releases/tag/3.0.0) which uses a direct WebSocket protocol for its subscription connection, and removes the MQTT stack (and old SocketRocket WebSocket client) entirely. Upgrading to that client version will let you automatically use the enhanced AppSync real time [pure WebSockets endpoints](https://aws.amazon.com/blogs/mobile/appsync-realtime/) without any configuration changes on your part.

Do note that the upgrade to 3.0.0 is a breaking change: the behavior of subscriptions that did not request full selection sets changed between 2.15.0 and 3.0.0, so it may require you to modify your client code and/or GraphQL subscriptions/mutations if you were relying on the old behavior. Please see the [AppSync 3.0.0 release notes](https://github.com/awslabs/aws-mobile-appsync-sdk-ios/releases/tag/3.0.0) for a more detailed discussion.

**Comment by adamup928** [comment]
Thanks, @palpatim - I'll close this issue and open a new one, if needed, given the significant changes.  I appreciate your help!
